lib/tbot/cli/globals.go

```diff
@@ -0,0 +1,114 @@
+	// - Use FIPS relevant endpoints for cloud providers (e.g AWS)
+	// - Restrict TLS / SSH cipher suites and TLS version
+	// - RSA2048 or ECDSA with NIST-P256 curve should be used for private key generation
+	FIPS bool
+
+	// ConfigPath is a path to a YAML configuration file to load, if any.
+	ConfigPath string
+
+	// Debug enables debug-level logging, when set
+	Debug bool
+
+	// LogFormat configures the output format of the logger
+	LogFormat string
+
+	// Trace indicates whether tracing should be enabled.
+	Trace bool
+
+	// TraceExporter is a manually provided URI to send traces to instead of
+	// forwarding them to the Auth service.
+	TraceExporter string
+
+	// Insecure instructs `tbot` to trust the Auth Server without verifying the CA.
+	Insecure bool
+
+	// staticConfigYAML allows tests to specify a configuration file statically
+	staticConfigYAML string
+}
+
+// NewGlobalArgs appends global flags to the application and returns a struct
+// that will be populated at parse time.
+func NewGlobalArgs(app *kingpin.Application) *GlobalArgs {
+	g := &GlobalArgs{}
+
+	app.Flag("debug", "Verbose logging to stdout.").Short('d').BoolVar(&g.Debug)
+	app.Flag("config", "Path to a configuration file.").Short('c').StringVar(&g.ConfigPath)
+	app.Flag("fips", "Runs tbot in FIPS compliance mode. This requires the FIPS binary is in use.").BoolVar(&g.FIPS)
+	app.Flag("trace", "Capture and export distributed traces.").Hidden().BoolVar(&g.Trace)
+	app.Flag("trace-exporter", "An OTLP exporter URL to send spans to.").Hidden().StringVar(&g.TraceExporter)
+	app.Flag("insecure", "Insecure configures the bot to trust the certificates from the Auth Server or Proxy on first connect without verification. Do not use in production.").BoolVar(&g.Insecure)
+	app.Flag("log-format", "Controls the format of output logs. Can be `json` or `text`. Defaults to `text`.").
+		Default(utils.LogFormatText).
+		EnumVar(&g.LogFormat, utils.LogFormatJSON, utils.LogFormatText)
+
+	return g
+}
+
+// NewGlobalArgsWithStaticConfig creates a new GlobalArgs instance with a static
+// YAML config. This can be used in tests to preload a config file without
+// writing to the filesystem. Note that this only works for codepaths that
+// make use of `LoadConfigWithMutators`.
+func NewGlobalArgsWithStaticConfig(staticYAML string) *GlobalArgs {
+	return &GlobalArgs{
+		staticConfigYAML: staticYAML,
+	}
+}
+
+func (g *GlobalArgs) ApplyConfig(cfg *config.BotConfig, l *slog.Logger) error {
+	// Note: g.ConfigPath is not checked here; the config must have already been
+	// loaded.
+
+	if g.FIPS {
+		cfg.FIPS = g.FIPS
+	}
+
+	if g.Debug {
+		cfg.Debug = g.Debug
+	}
+
+	// TODO: in previous versions, `insecure` is handled _after_
+	// BotConfig.CheckAndSetDefaults(). This flag is checked and setting it here
+	// *will* cause a behavioral change, so make sure the new behavior is sane.
+	// (It is unclear why this was done.)
+	if g.Insecure {
+		cfg.Insecure = true
+	}
```

Comment:
I'm a bit inclined to believe this old behavior was a bug and would allow `--ca-pins` and `--ca-path` to be set alongside `--insecure` which we try to consider an error in `BotConfig.CheckAndSetDefaults()`.

Reply:
It definitely seems like a case of user confusion if someone is specifying these flags together and I'd definitely rather explicitly block it. In my experience, gracefully handling invalid configurations only leads to more user confusion as they start to build up a mental model.

Reply:
👍 I'll keep the behavior as written here and remove the TODO. I figure this slight behavior change is one more vote in the "don't backport to v16, at least not yet" bucket.